pkg/kv/kvserver/raftlog/encoding.go

```diff
@@ -23,83 +23,128 @@ import (
 // and, in some cases, the first byte of the Entry's Data payload.
 type EntryEncoding byte
 
-// TODO(tbg): use auto-assigned consts (iota) for the encodings below, since
-// they aren't on the wire.
-
 const (
-	// EntryEncodingStandard is the default encoding for a CockroachDB raft log
-	// entry.
+	// EntryEncodingEmpty is an empty entry. These are used by raft after
+	// leader election. Since they hold no data, there is nothing in them to
+	// decode.
+	EntryEncodingEmpty EntryEncoding = iota
+	// EntryEncodingStandardWithAC is the default encoding for a CockroachDB
+	// raft log entry.
 	//
-	// This is a raftpb.Entry of type EntryNormal whose Data slice is either empty
-	// or whose first byte matches EntryEncodingStandardPrefixByte. The subsequent
-	// eight bytes represent a CmdIDKey. The remaining bytes represent a
-	// kvserverpb.RaftCommand.
-	EntryEncodingStandard EntryEncoding = 0
-	// EntryEncodingSideloaded indicates a proposal representing the result of a
-	// roachpb.AddSSTableRequest for which the payload (the SST) is stored outside
-	// the storage engine to improve storage performance.
+	// This is a raftpb.Entry of type EntryNormal whose Data slice is either
+	// empty or whose first byte matches entryEncodingStandardWithACPrefixByte.
+	// The subsequent eight bytes represent a CmdIDKey. The remaining bytes
+	// represent a kvserverpb.RaftCommand that also includes below-raft
+	// admission control metadata.
+	EntryEncodingStandardWithAC
+	// EntryEncodingSideloadedWithAC indicates a proposal representing the
+	// result of a roachpb.AddSSTableRequest for which the payload (the SST) is
+	// stored outside the storage engine to improve storage performance.
 	//
-	// This is a raftpb.Entry of type EntryNormal whose data slice is either empty
-	// or whose first byte matches EntryEncodingSideloadedPrefixByte. The subsequent
-	// eight bytes represent a CmdIDKey. The remaining bytes represent a
+	// This is a raftpb.Entry of type EntryNormal whose data slice is either
+	// empty or whose first byte matches
+	// entryEncodingSideloadedWithACPrefixByte. The subsequent eight bytes
+	// represent a CmdIDKey. The remaining bytes represent a
 	// kvserverpb.RaftCommand whose kvserverpb.ReplicatedEvalResult holds a
 	// nontrival kvserverpb.ReplicatedEvalResult_AddSSTable, the Data field of
 	// which is an SST to be ingested (and which is present in memory but made
 	// durable via direct storage on the filesystem, bypassing the storage
-	// engine).
-	EntryEncodingSideloaded EntryEncoding = 1
-	// EntryEncodingEmpty is an empty entry. These are used by raft after
-	// leader election. Since they hold no data, there is nothing in them to
-	// decode.
-	EntryEncodingEmpty EntryEncoding = 253
+	// engine). The kvserverpb.RaftCommand also includes below-raft admission
```

Comment:
"Includes" in which way? Let's make it precise.

Reply:
Ah, it just literally is stuff in the proto? Then maybe make that clear, I somehow thought you had stuffed it in there without the need to unmarshal the entire RaftCommand. Isn't it awkward to unmarshal the command? Or am I missing how this works? I've reviewed the entire PR at this point but still not sure.

Reply:
>  I somehow thought you had stuffed it in there without the need to unmarshal the entire RaftCommand. 

I did, but when separating out this PR this part stayed within #95748. Should've moved this part in here, sorry.

> Isn't it awkward to unmarshal the command? Or am I missing how this works?

#95748 adds the following, notice the proto tags -- they parallel the fields we added to kvserverpb.RaftCommand. So we can unmarshal selectively.

```proto
 // RaftAdmissionMeta contains information used by admission control for the
 // select raft commands that use replication admission control. It contains a
 // subset of the fields in kvserverpb.RaftCommand to selectively decode
 // state[1]. When marshaling a RaftCommand, we willfully include this data in
 // the prefix of the marshaled byte buffer. Information about whether this data
 // is present is captured in the first byte of the encoded raft proposal -- see
 // raftlog.EntryEncoding.
 //
 // [1]: The field tags and types must be kept identical with what's found there.
 message RaftAdmissionMeta {
   // AdmissionPriority of the command (maps to admission.WorkPriority); used
   // within a tenant below-raft for replication admission control.
   int32 admission_priority = 18;
   // AdmissionCreateTime is equivalent to Time.UnixNano() at the creation time
   // of the request, or a parent request, for which this command is a part of.
   // It's used within a tenant below-raft for replication admission control; see
   // admission.WorkInfo.CreateTime for details.
   int64 admission_create_time = 19;
   // AdmissionOriginNode captures where this raft command originated. It's used
   // to inform said node of this raft command's (virtual) admission in order for
   // it to release flow tokens for subsequent commands.
   int32 admission_origin_node = 20 [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"];
 }
```

```go
 // DecodeRaftAdmissionMeta decodes admission control metadata from a
 // raftpb.Entry.Data. Expects an EntryEncoding{Standard,Sideloaded}WithAC
 // encoding.
 func DecodeRaftAdmissionMeta(data []byte) (kvflowcontrolpb.RaftAdmissionMeta, error) {
 	prefix := data[0]
 	if !(prefix == entryEncodingStandardWithACPrefixByte || prefix == entryEncodingSideloadedWithACPrefixByte) {
 		panic(fmt.Sprintf("invalid encoding: prefix %v", prefix))
 	}

 	// TODO(irfansharif): If the decoding overhead is noticeable, we can write a
 	// custom decoder and rely on the encoding for raft admission data being
 	// present at the start of the marshaled raft command. This could speed it
 	// up slightly.
 	var raftAdmissionMeta kvflowcontrolpb.RaftAdmissionMeta
 	if err := protoutil.Unmarshal(data[1+RaftCommandIDLen:], &raftAdmissionMeta); err != nil {
 		return kvflowcontrolpb.RaftAdmissionMeta{}, err
 	}
 	return raftAdmissionMeta, nil
 }
```

Reply:
> they parallel the fields we added to kvserverpb.RaftCommand. So we can unmarshal selectively.

Ok, and that's good enough? Because to unmarshal selectively you'll still end up scanning the entire buffer, no?

Reply:
Yes, we do end up scanning the entire buffer. I ran and end-to-end benchmark earlier and the difference was in the noise. From [here](https://github.com/cockroachdb/cockroach/pull/93102/files#diff-4b1c6a6a37bd89d2660d1d36a1841335f255941e816597984e65ed6d1fa279baR721-R740):

```
// Using benchstat for before/after replication admission control proto changes:
 //
 //   ۩ dev bench pkg/kv/kvserver \
 //  	--filter BenchmarkReplicaProposal/bytes=256_B,withFollower=true \
 // 		-v --stream-output --ignore-cache --bench-time=5s --count 5
 //   goos: darwin
 //   goarch: arm64
 //   ...
 //
 //   name                                              old time/op    new time/op    delta
 //   ReplicaProposal/bytes=256_B,withFollower=true-10    45.4µs ± 2%    45.1µs ± 1%   ~     (p=1.000 n=5+5)
 //
 //   name                                              old speed      new speed      delta
 //   ReplicaProposal/bytes=256_B,withFollower=true-10  5.65MB/s ± 3%  5.67MB/s ± 1%   ~     (p=0.905 n=5+5)
 //
 //   name                                              old alloc/op   new alloc/op   delta
 //   ReplicaProposal/bytes=256_B,withFollower=true-10    26.3kB ± 1%    26.6kB ± 1%   ~     (p=0.056 n=5+5)
 //
 //   name                                              old allocs/op  new allocs/op  delta
 //   ReplicaProposal/bytes=256_B,withFollower=true-10       129 ± 1%       130 ± 1%   ~     (p=0.206 n=5+5)
```

This encoding/partial-decoding is noticeable in a micro-benchmark, from [here](https://github.com/cockroachdb/cockroach/pull/93102/files#diff-f3b4f34376eb9d955be6bc9f54bee3cf70cc80ee692f520d0ebb90b4cba51575R26-R38):

```
 BenchmarkRaftAdmissionMetaOverhead measures the overhead of encoding/decoding
 // raft metadata, compared to not doing it. It's structured similar to how raft
 // command data is encoded + decoded end-to-end, including the optional
 // below-raft admission control data, where steps (2) and (4) below are only
 // done for proposals subject to below-raft admission.
 //
 //	name                                                old time/op  new time/op  delta
 //	RaftAdmissionMetaOverhead/bytes=1.0_KiB,raft-ac-10  1.30µs ± 1%  1.70µs ± 1%  +30.43%  (p=0.008 n=5+5)
 //	RaftAdmissionMetaOverhead/bytes=256_KiB,raft-ac-10  51.6µs ± 4%  50.6µs ± 5%     ~     (p=0.421 n=5+5)
 //	RaftAdmissionMetaOverhead/bytes=512_KiB,raft-ac-10  91.9µs ± 4%  91.2µs ± 5%     ~     (p=1.000 n=5+5)
 //	RaftAdmissionMetaOverhead/bytes=1.0_MiB,raft-ac-10   148µs ± 4%   151µs ± 5%     ~     (p=0.095 n=5+5)
 //	RaftAdmissionMetaOverhead/bytes=2.0_MiB,raft-ac-10   290µs ± 3%   292µs ± 1%     ~     (p=0.151 n=5+5)
```

When encoding this meta, we're explicitly encoding it at the start of the marshaled buffer. So in theory we could write our own marshaler that bails early once it's encountered all the proto wire tags it's looking for. Once more of these pieces land and we have more rigorous benchmarks, we could implement that. There's a TODO in the code.